File: ecom-py/video_compressor.py
"""
Video Compressor - Background compression worker using FFmpeg
"""
import os
import subprocess
import threading
import queue
import logging
import shutil
import time
from typing import Dict, Optional, Callable, Tuple
from dataclasses import dataclass
from enum import Enum

logger = logging.getLogger(__name__)

# Windows process priority flags
PRIORITY_FLAGS = {
    'low': 0x00000040,         # IDLE_PRIORITY_CLASS
    'below_normal': 0x00004000, # BELOW_NORMAL_PRIORITY_CLASS
    'normal': 0x00000020        # NORMAL_PRIORITY_CLASS
}


class CompressionStatus(Enum):
    """Status of compression jobs."""
    PENDING = "pending"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"
    SKIPPED = "skipped"


@dataclass
class CompressionJob:
    """Represents a compression job in the queue."""
    video_path: str
    transaction_id: int
    settings: Dict
    output_path: Optional[str] = None


class VideoCompressor:
    """Background video compression worker using FFmpeg."""

    def __init__(self, on_complete: Optional[Callable] = None):
        """
        Initialize the video compressor.

        Args:
            on_complete: Callback function called when compression completes.
                        Signature: on_complete(transaction_id, success, result_data)
        """
        self._queue = queue.Queue()
        self._worker_thread: Optional[threading.Thread] = None
        self._running = False
        self._current_job: Optional[CompressionJob] = None
        self._on_complete = on_complete
        self._ffmpeg_path: Optional[str] = None
        self._ffmpeg_available: Optional[bool] = None
        self._lock = threading.Lock()

    def start(self):
        """Start the background compression worker."""
        if self._running:
            return

        self._running = True
        self._worker_thread = threading.Thread(target=self._worker_loop, daemon=True)
        self._worker_thread.start()
        logger.info("Video compressor worker started")

    def stop(self):
        """Stop the background compression worker gracefully."""
        self._running = False
        # Add None to queue to unblock worker
        self._queue.put(None)

        if self._worker_thread and self._worker_thread.is_alive():
            self._worker_thread.join(timeout=5.0)

        logger.info("Video compressor worker stopped")

    def check_ffmpeg_installed(self) -> Tuple[bool, str]:
        """
        Check if FFmpeg is installed and accessible.

        Returns:
            Tuple of (is_available, message)
        """
        if self._ffmpeg_available is not None:
            if self._ffmpeg_available:
                return True, f"FFmpeg found at: {self._ffmpeg_path}"
            else:
                return False, "FFmpeg not found. Please install FFmpeg to enable compression."

        # Try to find ffmpeg
        ffmpeg_cmd = "ffmpeg"

        # Get the directory where this script is located
        script_dir = os.path.dirname(os.path.abspath(__file__))

        # Check common locations on Windows
        if os.name == 'nt':
            common_paths = [
                # Check local project folder first
                os.path.join(script_dir, "ffmpeg.exe"),
                os.path.join(script_dir, "ffmpeg", "ffmpeg.exe"),
                os.path.join(script_dir, "ffmpeg", "bin", "ffmpeg.exe"),
                # Then check system paths
                r"C:\ffmpeg\bin\ffmpeg.exe",
                r"C:\Program Files\ffmpeg\bin\ffmpeg.exe",
                r"C:\Program Files (x86)\ffmpeg\bin\ffmpeg.exe",
                os.path.expanduser(r"~\ffmpeg\bin\ffmpeg.exe"),
            ]

            for path in common_paths:
                if os.path.isfile(path):
                    ffmpeg_cmd = path
                    logger.info(f"Found FFmpeg at: {path}")
                    break

        try:
            # Set up flags to hide console window on Windows
            creationflags = 0
            startupinfo = None
            if os.name == 'nt':
                creationflags = 0x08000000  # CREATE_NO_WINDOW
                startupinfo = subprocess.STARTUPINFO()
                startupinfo.dwFlags |= subprocess.STARTF_USESHOWWINDOW
                startupinfo.wShowWindow = subprocess.SW_HIDE

            result = subprocess.run(
                [ffmpeg_cmd, "-version"],
                capture_output=True,
                text=True,
                timeout=10,
                creationflags=creationflags,
                startupinfo=startupinfo
            )

            if result.returncode == 0:
                self._ffmpeg_available = True
                self._ffmpeg_path = ffmpeg_cmd
                version_line = result.stdout.split('\n')[0] if result.stdout else "Unknown version"
                logger.info(f"FFmpeg found: {version_line}")
                return True, f"FFmpeg available: {version_line}"
            else:
                self._ffmpeg_available = False
                return False, "FFmpeg found but returned an error"

        except FileNotFoundError:
            self._ffmpeg_available = False
            return False, "FFmpeg not found. Please install FFmpeg and add it to your PATH."
        except subprocess.TimeoutExpired:
            self._ffmpeg_available = False
            return False, "FFmpeg check timed out"
        except Exception as e:
            self._ffmpeg_available = False
            return False, f"Error checking FFmpeg: {str(e)}"

    def queue_compression(
        self,
        video_path: str,
        transaction_id: int,
        settings: Dict
    ) -> bool:
        """
        Queue a video for compression.

        Args:
            video_path: Path to the video file to compress
            transaction_id: Database transaction ID for status updates
            settings: Compression settings dict with keys:
                     - enabled: bool
                     - codec: 'h264' or 'h265'
                     - crf: int (18-35)
                     - preset: str (ultrafast/fast/medium/slow)
                     - delete_original: bool

        Returns:
            True if job was queued, False otherwise
        """
        if not settings.get('enabled', False):
            logger.info(f"Compression disabled, skipping {video_path}")
            if self._on_complete:
                self._on_complete(transaction_id, True, {
                    'status': CompressionStatus.SKIPPED.value,
                    'message': 'Compression disabled'
                })
            return False

        # Check if ffmpeg is available
        available, _ = self.check_ffmpeg_installed()
        if not available:
            logger.warning(f"FFmpeg not available, skipping compression for {video_path}")
            if self._on_complete:
                self._on_complete(transaction_id, False, {
                    'status': CompressionStatus.SKIPPED.value,
                    'message': 'FFmpeg not installed'
                })
            return False

        # Validate video file exists
        if not os.path.isfile(video_path):
            logger.error(f"Video file not found: {video_path}")
            if self._on_complete:
                self._on_complete(transaction_id, False, {
                    'status': CompressionStatus.FAILED.value,
                    'message': 'Video file not found'
                })
            return False

        job = CompressionJob(
            video_path=video_path,
            transaction_id=transaction_id,
            settings=settings
        )

        self._queue.put(job)
        logger.info(f"Queued compression job for transaction {transaction_id}: {video_path}")
        return True

    def get_queue_status(self) -> Dict:
        """
        Get the current status of the compression queue.

        Returns:
            Dict with queue status information
        """
        with self._lock:
            return {
                'queue_size': self._queue.qsize(),
                'is_processing': self._current_job is not None,
                'current_transaction_id': self._current_job.transaction_id if self._current_job else None,
                'worker_running': self._running
            }

    def _worker_loop(self):
        """Main worker loop that processes compression jobs."""
        logger.info("Compression worker loop started")

        while self._running:
            try:
                # Wait for a job with timeout
                job = self._queue.get(timeout=1.0)

                if job is None:
                    # Shutdown signal
                    break

                with self._lock:
                    self._current_job = job

                # Process the job
                self._process_job(job)

                with self._lock:
                    self._current_job = None

                self._queue.task_done()

            except queue.Empty:
                continue
            except Exception as e:
                logger.error(f"Error in compression worker: {e}")
                with self._lock:
                    self._current_job = None

        logger.info("Compression worker loop ended")

    def _process_job(self, job: CompressionJob):
        """
        Process a single compression job.

        Args:
            job: The compression job to process
        """
        logger.info(f"Processing compression job for transaction {job.transaction_id}")

        try:
            # Generate output path
            base_path, ext = os.path.splitext(job.video_path)
            output_path = f"{base_path}_compressed{ext}"
            job.output_path = output_path

            # Get original file size
            original_size = os.path.getsize(job.video_path)
            original_size_mb = original_size / (1024 * 1024)

            # Run compression
            success, message = self._compress_video(job)

            if success and os.path.isfile(output_path):
                # Get compressed file size
                compressed_size = os.path.getsize(output_path)
                compressed_size_mb = compressed_size / (1024 * 1024)
                compression_ratio = (1 - (compressed_size / original_size)) * 100 if original_size > 0 else 0

                logger.info(
                    f"Compression complete: {original_size_mb:.2f}MB -> {compressed_size_mb:.2f}MB "
                    f"({compression_ratio:.1f}% reduction)"
                )

                # Handle original file
                compressed_filename = os.path.basename(output_path)
                if job.settings.get('delete_original', True):
                    try:
                        os.remove(job.video_path)
                        # Rename compressed file to original name
                        os.rename(output_path, job.video_path)
                        compressed_filename = os.path.basename(job.video_path)
                        logger.info(f"Replaced original file with compressed version")
                    except Exception as e:
                        logger.error(f"Error replacing original file: {e}")

                # Callback with success
                if self._on_complete:
                    self._on_complete(job.transaction_id, True, {
                        'status': CompressionStatus.COMPLETED.value,
                        'compressed_file_size_mb': compressed_size_mb,
                        'compression_ratio': compression_ratio,
                        'compressed_filename': compressed_filename,
                        'message': f'Compressed successfully ({compression_ratio:.1f}% reduction)'
                    })
            else:
                # Compression failed
                logger.error(f"Compression failed: {message}")

                # Clean up partial output if exists
                if os.path.isfile(output_path):
                    try:
                        os.remove(output_path)
                    except:
                        pass

                if self._on_complete:
                    self._on_complete(job.transaction_id, False, {
                        'status': CompressionStatus.FAILED.value,
                        'message': message
                    })

        except Exception as e:
            logger.error(f"Error processing compression job: {e}")
            if self._on_complete:
                self._on_complete(job.transaction_id, False, {
                    'status': CompressionStatus.FAILED.value,
                    'message': str(e)
                })

    def _compress_video(self, job: CompressionJob) -> Tuple[bool, str]:
        """
        Execute FFmpeg compression command with low priority to avoid blocking main app.

        Args:
            job: The compression job with settings

        Returns:
            Tuple of (success, message)
        """
        settings = job.settings
        codec = settings.get('codec', 'h264')
        crf = settings.get('crf', 23)
        preset = settings.get('preset', 'medium')
        priority = settings.get('priority', 'below_normal')

        # Build FFmpeg command
        ffmpeg_cmd = self._ffmpeg_path or 'ffmpeg'

        # Select video codec
        if codec == 'h265':
            video_codec = 'libx265'
        else:
            video_codec = 'libx264'

        # Use faster preset and limit CPU threads to reduce impact on main app
        cmd = [
            ffmpeg_cmd,
            '-i', job.video_path,
            '-c:v', video_codec,
            '-crf', str(crf),
            '-preset', preset,
            '-threads', '2',  # Limit CPU threads
            '-c:a', 'aac',
            '-b:a', '128k',
            '-y',  # Overwrite output
            job.output_path
        ]

        logger.info(f"Running FFmpeg command (priority: {priority}): {' '.join(cmd)}")
        logger.info(f"Input file: {job.video_path} (exists: {os.path.isfile(job.video_path)})")
        logger.info(f"Output file: {job.output_path}")

        try:
            # Set up process creation flags for Windows
            creationflags = 0
            startupinfo = None
            if os.name == 'nt':
                # CREATE_NO_WINDOW = 0x08000000 (hide console window)
                priority_flag = PRIORITY_FLAGS.get(priority, 0x00004000)
                creationflags = 0x08000000 | priority_flag

                # Also use startupinfo to ensure no window
                startupinfo = subprocess.STARTUPINFO()
                startupinfo.dwFlags |= subprocess.STARTF_USESHOWWINDOW
                startupinfo.wShowWindow = subprocess.SW_HIDE

            # Use Popen for non-blocking execution with low priority and no window
            # Redirect all output to DEVNULL to prevent pipe buffer blocking
            process = subprocess.Popen(
                cmd,
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL,
                stdin=subprocess.DEVNULL,
                creationflags=creationflags,
                startupinfo=startupinfo
            )

            # Wait for completion with timeout (non-blocking check every 2 seconds)
            timeout_seconds = 3600  # 1 hour
            elapsed = 0
            last_log = 0
            while elapsed < timeout_seconds:
                retcode = process.poll()
                if retcode is not None:
                    break
                time.sleep(2)
                elapsed += 2
                # Log progress every 30 seconds
                if elapsed - last_log >= 30:
                    logger.info(f"Compression in progress... ({elapsed}s elapsed)")
                    last_log = elapsed

            if process.poll() is None:
                # Process still running after timeout, kill it
                process.kill()
                process.wait()
                return False, "Compression timed out (exceeded 1 hour)"

            # Get the return code
            returncode = process.returncode
            logger.info(f"FFmpeg process completed with return code: {returncode}")

            if returncode == 0:
                return True, "Compression successful"
            else:
                logger.error(f"FFmpeg failed with return code {returncode}")
                return False, f"FFmpeg error (return code {returncode})"

        except Exception as e:
            return False, f"Error running FFmpeg: {str(e)}"

File: ecom-flutter/lib/core/video_compressor.dart
/// Video Compressor - Background compression worker using FFmpeg.
///
/// Behavioral port of ecom-py/video_compressor.py's queue.Queue +
/// daemon-thread design onto Dart's single-threaded event loop: a FIFO
/// queue drained by a strictly-serialized async pump (one job in flight at
/// a time), with completion delivered through an injected callback rather
/// than a return value - the caller never blocks (COMP-01).
///
/// One deliberate divergence from ecom-py (COMP-03): when replacing the
/// original, the swap goes original -> .bak, temp -> original, delete
/// .bak, so every crash point leaves at least one intact copy of the
/// video on disk (ecom-py deletes the original before renaming).
library;

import 'dart:async';
import 'dart:collection';
import 'dart:io';

import 'package:flutter/foundation.dart' show ValueNotifier;
import 'package:path/path.dart' as p;

import 'config.dart';
import 'ffmpeg_locator.dart';
import 'logger.dart';
import 'process_priority.dart';

/// Status of compression jobs. [value] strings are what's persisted to the
/// transactions table - identical to ecom-py's CompressionStatus enum.
enum CompressionStatus {
  pending('pending'),
  processing('processing'),
  completed('completed'),
  failed('failed'),
  skipped('skipped');

  const CompressionStatus(this.value);

  /// The database/string form of the status (matches ecom-py rows).
  final String value;
}

/// Compression parameters for a single job - mirrors the settings dict
/// ecom-py passes into queue_compression().
class CompressionSettings {
  const CompressionSettings({
    this.enabled = true,
    this.codec = 'h264',
    this.crf = 23,
    this.preset = 'medium',
    this.deleteOriginal = true,
    this.priority = 'below_normal',
  });

  /// Builds settings from the app [Config]'s `compression` category.
  factory CompressionSettings.fromConfig(Config config) {
    return CompressionSettings(
      enabled: config.compressionEnabled,
      codec: config.compressionCodec,
      crf: config.compressionCrf,
      preset: config.compressionPreset,
      deleteOriginal: config.compressionDeleteOriginal,
      priority: config.compressionPriority,
    );
  }

  /// Master enable toggle; disabled jobs are skipped with a callback.
  final bool enabled;

  /// 'h264' (libx264) or 'h265' (libx265).
  final String codec;

  /// Constant Rate Factor; clamped to 18-35 at command construction.
  final int crf;

  /// FFmpeg preset (ultrafast/fast/medium/slow).
  final String preset;

  /// Replace the original with the compressed file (true, ecom-py
  /// default) or keep both files side by side (false).
  final bool deleteOriginal;

  /// Windows process priority: 'low', 'below_normal', or 'normal'.
  final String priority;
}

/// Represents a compression job in the queue (ecom-py CompressionJob).
class CompressionJob {
  CompressionJob({
    required this.videoPath,
    required this.transactionId,
    required this.settings,
    this.videoFilter,
  });

  final String videoPath;
  final int transactionId;
  final CompressionSettings settings;

  /// Optional -vf filtergraph applied during the encode (the watermark
  /// burn-in). Merging the watermark into the compression pass means every
  /// recording is re-encoded exactly once.
  final String? videoFilter;
  String? outputPath;
}

/// Snapshot of the compressor's queue state, published via
/// [VideoCompressor.queueStatus] - field-for-field ecom-py's
/// get_queue_status() dict.
class CompressionQueueStatus {
  const CompressionQueueStatus({
    required this.queueSize,
    required this.isProcessing,
    this.currentTransactionId,
    required this.workerRunning,
  });

  final int queueSize;
  final bool isProcessing;
  final int? currentTransactionId;
  final bool workerRunning;

  @override
  bool operator ==(Object other) {
    return other is CompressionQueueStatus &&
        other.queueSize == queueSize &&
        other.isProcessing == isProcessing &&
        other.currentTransactionId == currentTransactionId &&
        other.workerRunning == workerRunning;
  }

  @override
  int get hashCode =>
      Object.hash(queueSize, isProcessing, currentTransactionId, workerRunning);

  @override
  String toString() =>
      'CompressionQueueStatus(queueSize: $queueSize, isProcessing: '
      '$isProcessing, currentTransactionId: $currentTransactionId, '
      'workerRunning: $workerRunning)';
}

/// Callback invoked when a job finishes (or is skipped/failed before
/// queuing). [resultData] keys match ecom-py's on_complete dict: 'status',
/// 'compressed_file_size_mb', 'compression_ratio', 'compressed_filename',
/// 'message'.
typedef CompressionCompleteCallback = void Function(
  int transactionId,
  bool success,
  Map<String, Object?> resultData,
);

/// Seam for spawning the FFmpeg process (fake-able in tests).
typedef ProcessStarter = Future<Process> Function(
  String executable,
  List<String> arguments,
);

/// Seam for FFmpeg discovery (defaults to [FfmpegLocator.findFfmpeg]).
typedef FfmpegFinder = Future<String?> Function();

/// Seam for probing a video's source bitrate in kbit/s (null when it cannot
/// be determined). Defaults to an ffprobe invocation.
typedef SourceBitrateProber = Future<int?> Function(String videoPath);

/// Seam for applying process priority (defaults to [setProcessPriority]).
typedef PriorityApplier = bool Function(int pid, String priority);

/// Default [SourceBitrateProber]: asks ffprobe for the video stream's
/// bitrate (falling back to the container bitrate). Returns kbit/s, or null
/// when ffprobe is missing or the probe fails - callers then encode without
/// a rate cap, exactly the pre-probe behavior.
Future<int?> probeSourceBitrateKbps(String videoPath) async {
  final ffprobe = await FfmpegLocator.findFfprobe();
  if (ffprobe == null) {
    return null;
  }
  try {
    final result = await Process.run(ffprobe, [
      '-v', 'error',
      '-select_streams', 'v:0',
      '-show_entries', 'stream=bit_rate:format=bit_rate',
      '-of', 'default=noprint_wrappers=1:nokey=1',
      videoPath,
    ]);
    if (result.exitCode != 0) {
      return null;
    }
    // Stream bitrate is printed first when known; some containers only
    // report a format-level bitrate ("N/A" lines are skipped).
    for (final line in '${result.stdout}'.split(RegExp(r'\r?\n'))) {
      final bps = int.tryParse(line.trim());
      if (bps != null && bps > 0) {
        return bps ~/ 1000;
      }
    }
  } catch (_) {
    // Probe failures are non-fatal - the encode just runs uncapped.
  }
  return null;
}

/// Background video compression worker using FFmpeg.
class VideoCompressor {
  /// [onComplete] is called for every terminal outcome (completed, failed,
  /// skipped). The remaining parameters are injectable seams so queue
  /// serialization, command construction, atomic swap, timeout, and
  /// graceful-skip behavior are all unit-testable without FFmpeg
  /// installed.
  VideoCompressor(
    this._logger, {
    CompressionCompleteCallback? onComplete,
    ProcessStarter? processStarter,
    FfmpegFinder? ffmpegFinder,
    PriorityApplier? priorityApplier,
    SourceBitrateProber? bitrateProber,
    Duration jobTimeout = const Duration(hours: 1),
  })  : _onComplete = onComplete,
        _processStarter = processStarter ?? _defaultProcessStarter,
        _ffmpegFinder = ffmpegFinder ?? FfmpegLocator.findFfmpeg,
        _priorityApplier = priorityApplier ?? setProcessPriority,
        _bitrateProber = bitrateProber ?? probeSourceBitrateKbps,
        _jobTimeout = jobTimeout;

  final Logger _logger;
  final CompressionCompleteCallback? _onComplete;
  final ProcessStarter _processStarter;
  final FfmpegFinder _ffmpegFinder;
  final PriorityApplier _priorityApplier;
  final SourceBitrateProber _bitrateProber;
  final Duration _jobTimeout;

  final ListQueue<CompressionJob> _queue = ListQueue<CompressionJob>();
  bool _running = false;
  CompressionJob? _currentJob;
  Future<void>? _activeJob;
  String? _ffmpegPath;

  /// Live queue state for the UI (UI-01's compression queue readout).
  /// Phase 2's status panel consumes this via ValueListenableBuilder.
  final ValueNotifier<CompressionQueueStatus> queueStatus =
      ValueNotifier<CompressionQueueStatus>(
    const CompressionQueueStatus(
      queueSize: 0,
      isProcessing: false,
      workerRunning: false,
    ),
  );

  /// Jobs waiting plus the one in flight (0 when idle).
  int get queueLength => _queue.length + (_currentJob == null ? 0 : 1);

  /// Starts the background compression worker.
  void start() {
    if (_running) {
      return;
    }
    _running = true;
    _logger.info('Video compressor worker started');
    _notifyStatus();
    _pump();
  }

  /// Stops the worker gracefully: no new jobs start, and the returned
  /// future completes once the in-flight job (if any) finishes. Queued
  /// jobs remain in the queue.
  Future<void> stop() async {
    _running = false;
    _notifyStatus();
    final active = _activeJob;
    if (active != null) {
      await active;
    }
    _logger.info('Video compressor worker stopped');
  }

  /// Queues a video for compression. Returns true if the job was queued.
  ///
  /// Mirrors ecom-py's queue_compression() short-circuits: when
  /// compression is disabled or FFmpeg is unavailable the job is skipped
  /// gracefully (logged + 'skipped' callback, recording never affected);
  /// a missing input file reports 'failed'.
  ///
  /// [videoFilter] burns the watermark during the same encode. A job with a
  /// filter runs even when compression is disabled - the watermark requires
  /// a re-encode regardless, and encoding once at the configured CRF is
  /// strictly cheaper than a separate watermark pass. The enabled toggle
  /// still governs plain (filter-less) backlog jobs.
  Future<bool> queueCompression(
    String videoPath,
    int transactionId,
    CompressionSettings settings, {
    String? videoFilter,
  }) async {
    if (!settings.enabled && videoFilter == null) {
      _logger.info('Compression disabled, skipping $videoPath');
      _onComplete?.call(transactionId, true, {
        'status': CompressionStatus.skipped.value,
        'message': 'Compression disabled',
      });
      return false;
    }

    final ffmpegPath = await _ffmpegFinder();
    if (ffmpegPath == null) {
      _logger.warning(
        'FFmpeg not available, skipping compression for $videoPath',
      );
      _onComplete?.call(transactionId, false, {
        'status': CompressionStatus.skipped.value,
        'message': 'FFmpeg not installed',
      });
      return false;
    }
    _ffmpegPath = ffmpegPath;

    if (!File(videoPath).existsSync()) {
      _logger.error('Video file not found: $videoPath');
      _onComplete?.call(transactionId, false, {
        'status': CompressionStatus.failed.value,
        'message': 'Video file not found',
      });
      return false;
    }

    _queue.add(CompressionJob(
      videoPath: videoPath,
      transactionId: transactionId,
      settings: settings,
      videoFilter: videoFilter,
    ));
    _logger.info(
      'Queued compression job for transaction $transactionId: $videoPath',
    );
    _notifyStatus();
    _pump();
    return true;
  }

  /// Serialized FIFO pump: starts the next job only when the worker is
  /// running and nothing is in flight. Re-entered from each job's
  /// whenComplete, so jobs can never overlap (COMP-01).
  void _pump() {
    if (!_running || _currentJob != null || _queue.isEmpty) {
      return;
    }
    final job = _queue.removeFirst();
    _currentJob = job;
    _notifyStatus();
    _activeJob = _processJob(job).whenComplete(() {
      _currentJob = null;
      _activeJob = null;
      _notifyStatus();
      _pump();
    });
  }

  void _notifyStatus() {
    queueStatus.value = CompressionQueueStatus(
      queueSize: _queue.length,
      isProcessing: _currentJob != null,
      currentTransactionId: _currentJob?.transactionId,
      workerRunning: _running,
    );
  }

  /// Processes a single compression job (ecom-py _process_job).
  Future<void> _processJob(CompressionJob job) async {
    _logger.info(
      'Processing compression job for transaction ${job.transactionId}',
    );
    try {
      // Temp output lives in the SAME directory as the original so the
      // final rename is an atomic same-volume move (COMP-03).
      final ext = p.extension(job.videoPath);
      final base = job.videoPath.substring(0, job.videoPath.length - ext.length);
      final outputPath = '${base}_compressed$ext';
      job.outputPath = outputPath;

      final originalSize = File(job.videoPath).lengthSync();
      final originalSizeMb = originalSize / (1024 * 1024);

      final (success, message) = await _compressVideo(job);

      final outputFile = File(outputPath);
      final outputSize = outputFile.existsSync() ? outputFile.lengthSync() : 0;

      if (success && outputSize > 0 &&
          job.videoFilter == null && outputSize >= originalSize) {
        // The source was already leaner than what this CRF produces, so the
        // "compressed" file came out no smaller. Nothing but size is at
        // stake on a filter-less job - keep the original.
        try {
          outputFile.deleteSync();
        } catch (_) {
          // Best-effort cleanup.
        }
        _logger.info(
          'Compression not beneficial '
          '(${originalSizeMb.toStringAsFixed(2)}MB -> '
          '${(outputSize / (1024 * 1024)).toStringAsFixed(2)}MB), '
          'keeping original',
        );
        _onComplete?.call(job.transactionId, true, {
          'status': CompressionStatus.completed.value,
          'compressed_file_size_mb': originalSizeMb,
          'compression_ratio': 0.0,
          'compressed_filename': p.basename(job.videoPath),
          'message': 'Compression not beneficial - kept original',
        });
        return;
      }

      if (success && outputSize > 0) {
        final compressedSizeMb = outputSize / (1024 * 1024);
        final compressionRatio =
            originalSize > 0 ? (1 - outputSize / originalSize) * 100 : 0.0;
        _logger.info(
          'Compression complete: ${originalSizeMb.toStringAsFixed(2)}MB -> '
          '${compressedSizeMb.toStringAsFixed(2)}MB '
          '(${compressionRatio.toStringAsFixed(1)}% reduction)',
        );

        var compressedFilename = p.basename(outputPath);
        if (job.settings.deleteOriginal) {
          compressedFilename = _swapInCompressed(job, outputFile) ??
              compressedFilename;
        }
        // deleteOriginal=false keeps BOTH files (ecom-py parity) and
        // records the _compressed basename in the DB.

        _onComplete?.call(job.transactionId, true, {
          'status': CompressionStatus.completed.value,
          'compressed_file_size_mb': compressedSizeMb,
          'compression_ratio': compressionRatio,
          'compressed_filename': compressedFilename,
          'message': 'Compressed successfully '
              '(${compressionRatio.toStringAsFixed(1)}% reduction)',
        });
      } else {
        final failMessage =
            success ? 'FFmpeg produced an empty output file' : message;
        _logger.error('Compression failed: $failMessage');

        // Clean up partial output if it exists - the original is never
        // touched on the failure path.
        if (outputFile.existsSync()) {
          try {
            outputFile.deleteSync();
          } catch (_) {
            // Best-effort cleanup.
          }
        }

        _onComplete?.call(job.transactionId, false, {
          'status': CompressionStatus.failed.value,
          'message': failMessage,
        });
      }
    } catch (e) {
      _logger.error('Error processing compression job: $e');
      _onComplete?.call(job.transactionId, false, {
        'status': CompressionStatus.failed.value,
        'message': '$e',
      });
    }
  }

  /// Atomically replaces the original with the compressed file (COMP-03):
  /// original -> .bak, temp -> original, delete .bak. Every intermediate
  /// crash point leaves at least one intact copy on disk. Returns the
  /// final filename, or null if the swap failed (in which case a
  /// best-effort restore of the original is attempted).
  String? _swapInCompressed(CompressionJob job, File outputFile) {
    final backupPath = '${job.videoPath}.bak';
    try {
      File(job.videoPath).renameSync(backupPath);
      outputFile.renameSync(job.videoPath);
      File(backupPath).deleteSync();
      _logger.info('Replaced original file with compressed version');
      return p.basename(job.videoPath);
    } catch (e) {
      _logger.error('Error replacing original file: $e');
      // Restore the original if the swap left it missing.
      try {
        if (!File(job.videoPath).existsSync() &&
            File(backupPath).existsSync()) {
          File(backupPath).renameSync(job.videoPath);
        }
      } catch (restoreError) {
        _logger.error('Error restoring original file: $restoreError');
      }
      return null;
    }
  }

  /// Executes the FFmpeg command for [job]. Returns (success, message).
  Future<(bool, String)> _compressVideo(CompressionJob job) async {
    final settings = job.settings;
    final videoCodec = settings.codec == 'h265' ? 'libx265' : 'libx264';
    final crf = settings.crf.clamp(18, 35);

    // Cap the encode at the source's bitrate: recordings are already H.264
    // at the configured recording bitrate, and an uncapped CRF encode can
    // exceed it (e.g. CRF 23 at 4K produces ~9 Mbit/s - "compressing" a
    // 6 Mbit/s original into a larger file). null = probe unavailable,
    // encode uncapped as before.
    final sourceKbps = await _bitrateProber(job.videoPath);

    // Same invocation as ecom-py: capped threads so compression can't
    // starve capture/UI, AAC audio passthrough settings, -y to overwrite
    // a stale temp file.
    final args = <String>[
      '-i', job.videoPath,
      if (job.videoFilter != null) ...['-vf', job.videoFilter!],
      '-c:v', videoCodec,
      '-crf', '$crf',
      '-preset', settings.preset,
      if (sourceKbps != null && sourceKbps > 0) ...[
        '-maxrate', '${sourceKbps}k',
        '-bufsize', '${sourceKbps * 2}k',
      ],
      '-threads', '2',
      '-c:a', 'aac',
      '-b:a', '128k',
      '-y',
      job.outputPath!,
    ];

    final executable = _ffmpegPath ?? 'ffmpeg';
    _logger.info(
      'Running FFmpeg command (priority: ${settings.priority}): '
      '$executable ${args.join(' ')}',
    );

    try {
      final process = await _processStarter(executable, args);

      // Lower the child's priority immediately after spawn (SET-06).
      try {
        if (!_priorityApplier(process.pid, settings.priority)) {
          _logger.warning(
            'Could not set FFmpeg process priority (${settings.priority})',
          );
        }
      } catch (e) {
        _logger.warning('Error setting FFmpeg process priority: $e');
      }

      // Drain both pipes so FFmpeg can never block on a full pipe buffer
      // (ecom-py redirects to DEVNULL for the same reason).
      unawaited(process.stdout.drain<void>());
      unawaited(process.stderr.drain<void>());

      var timedOut = false;
      final exitCode = await process.exitCode.timeout(
        _jobTimeout,
        onTimeout: () {
          timedOut = true;
          process.kill(ProcessSignal.sigkill);
          // Give the kill a bounded window to take effect.
          return process.exitCode.timeout(
            const Duration(seconds: 10),
            onTimeout: () => -1,
          );
        },
      );

      if (timedOut) {
        return (
          false,
          'Compression timed out (exceeded ${_jobTimeout.inSeconds}s)',
        );
      }

      _logger.info('FFmpeg process completed with return code: $exitCode');
      if (exitCode == 0) {
        return (true, 'Compression successful');
      }
      return (false, 'FFmpeg error (return code $exitCode)');
    } catch (e) {
      return (false, 'Error running FFmpeg: $e');
    }
  }

  /// Default starter: argument-list spawn, never through a shell
  /// (T-03-01). Whether a console window flashes on real hardware is a
  /// Phase-5 packaging checkpoint (FFmpeg is not installed on the dev
  /// machine); if it does, swap this seam for an FFI CreateProcess with
  /// CREATE_NO_WINDOW.
  static Future<Process> _defaultProcessStarter(
    String executable,
    List<String> arguments,
  ) {
    return Process.start(executable, arguments, runInShell: false);
  }
}
